### ISSUE TITLE
Suggestion: auto-expansion of `<details>…</details>` for direct links

### ISSUE THREAD

**Issue by brianrourkeboll** [fsprojects/FSharp.Formatting#916]
While #778, as implemented in #818, does indeed make for easier skimming of API docs, I think it would be nice if:

- [x] There were some kind of "expand/collapse all" button.

   Yes, I can click expand for every member individually if I want to browse examples, or I can throw

   ```js
   for (const deets of document.getElementsByTagName('details')) {
       deets.setAttribute('open', 'true');
   }
   ```

    in the console, but it would be nice to have a button.

- [x] The `<details>…</details>` for a given member were auto-expanded when linked to directly.

   For example, linking directly to `List.tryPick` like https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-listmodule.html#tryPick currently shows this:

   ![image](https://github.com/fsprojects/FSharp.Formatting/assets/14795984/144a90fe-46c2-4bdb-a9cf-34cf4e098622)

   but I think it would make sense to auto-expand the details for the linked member:

   ![image](https://github.com/fsprojects/FSharp.Formatting/assets/14795984/1a40765c-2c93-4f35-a634-1a34d958355d)

(I'd be willing to open a PR for these.)

**Comment by nojaf** [comment]
Sounds reasonable, go for it!

**Comment by nojaf** [comment]
Hmm, still some remaining problems with the new approach:

https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-codeanalysis-fsharpchecker.html#InvalidateAll

![image](https://github.com/fsprojects/FSharp.Formatting/assets/2621499/ca3a8eb2-4b6f-4c75-be9b-fe75ff97923d)

**Comment by brianrourkeboll** [comment]
@nojaf Hmm, I'll look.

Separately, I'm noticing else on that same page that will necessitate some additional tweaks in that area anyway.

The `.fsdocs-entity-xmldoc { > div { flex-direction: row reverse } }`, which was already there, affects the order of any paragraphs defined with `<para>` that are nested in the `<summary>`.

So these [paragraphs](https://github.com/dotnet/fsharp/blob/2a65095c471dc557a5720d5c498d9ac18336649e/src/Compiler/Service/service.fsi#L178-L188) —

```fsharp
/// <summary>
/// <para>
///   Parse and check a source code file, returning a handle to the results
/// </para>
/// <para>
///    Note: all files except the one being checked are read from the FileSystem API
/// </para>
/// <para>
///   Return FSharpCheckFileAnswer.Aborted if a parse tree was not available.
/// </para>
/// </summary>
```

— are shown in reverse order as

![image](https://github.com/fsprojects/FSharp.Formatting/assets/14795984/6978a8f8-1022-4606-ba05-9687d14abaf8)

I think they should at least be shown in the order in which they're written, and it would probably be better if they were separated vertically instead of horizontally.

**Comment by nojaf** [comment]
Hmm, yeah, you're right. That row reverse was probably there for a reason. I'm okay with changing it, but we might need to consider any other issues it could introduce.

**Comment by brianrourkeboll** [comment]
Yeah, the contents themselves are generated in reverse order, only to be reversed again by the CSS, although I don't understand why:

https://github.com/fsprojects/FSharp.Formatting/blob/56b856bc3be5ab7ff7c1bbcb30fe0f3e758235cc/src/FSharp.Formatting.ApiDocs/GenerateHtml.fs#L183-L189

**Comment by brianrourkeboll** [comment]
> Hmm, still some remaining problems with the new approach:
> 
> https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-codeanalysis-fsharpchecker.html#InvalidateAll
> 
> ![image](https://private-user-images.githubusercontent.com/2621499/335492156-ca3a8eb2-4b6f-4c75-be9b-fe75ff97923d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcxNTUzMTAsIm5iZiI6MTcxNzE1NTAxMCwicGF0aCI6Ii8yNjIxNDk5LzMzNTQ5MjE1Ni1jYTNhOGViMi00YjZmLTRjNzUtYmU5Yi1mZTc1ZmY5NzkyM2QucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUzMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MzFUMTEzMDEwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzA5NjQ1ZjcyMDNlZWExZWVjNDE0NGI1NDUxNzJkM2Q0ZGIyMzQ0NzQ2NTQ3M2M4MjIwYzc3YjhjOTc5MTc1MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.GpfVj4s7EuVfi9k6SSSaSSlN0_TNWc8vXI6cve-jWfI)

Here's the explanation for this particular thing:

![image](https://github.com/fsprojects/FSharp.Formatting/assets/14795984/406e33bd-b4dc-4c6c-81de-8621f8d2654e)

Sometimes the `<div class="fsdocs-summary">` will be nested inside `<details><summary>`, and sometimes it won't be:

https://github.com/fsprojects/FSharp.Formatting/blob/56b856bc3be5ab7ff7c1bbcb30fe0f3e758235cc/src/FSharp.Formatting.ApiDocs/GenerateHtml.fs#L271-L277

That conditional actually also means that, if there is no `<summary>` in the XML doc, there will be no source links, [even if there are other elements](https://github.com/dotnet/fsharp/blob/e5d0212435b1c08bbc1fc72bcfbbbe7687a3f87b/src/Compiler/Service/service.fsi#L254-L269):

![image](https://github.com/fsprojects/FSharp.Formatting/assets/14795984/cb736b62-9119-4397-ac55-63829d2f8e2e)

I don't think there's a good reason not to show source links in such a case.

**Comment by brianrourkeboll** [comment]
If I did add an expand/collapse-all button, does this look like a reasonable place to put it? The idea would be for it to save your preference for the whole site to local storage, like the theme toggle.

https://github.com/fsprojects/FSharp.Formatting/assets/14795984/cef2961a-6619-445c-8dfe-2de4ec56dcde

I guess it could also only be shown on pages that had API docs, but then would it still make sense to show it in the top right? Or somewhere else?

**Comment by nojaf** [comment]
My 2 cents: it would only make sense on pages with API docs and should not be part of the menu bar.

**Comment by nhirschey** [comment]
*If it’s easy*, to me it seems more intuitive on the same row as “instance members”. For me it doesn’t seem obvious to look for this on the menu bar. I expect expand/collapse to be located closer to the elements it affects.

**Comment by brianrourkeboll** [comment]
@nhirschey

> _If it’s easy_, to me it seems more intuitive on the same row as “instance members”. For me it doesn’t seem obvious to look for this on the menu bar. I expect expand/collapse to be located closer to the elements it affects.

Something like this?

https://github.com/fsprojects/FSharp.Formatting/assets/14795984/040326cb-ca9e-4908-871a-5e0f9eeb9aed

**Comment by nhirschey** [comment]
> Something like this?

Yeah, exactly. To me that’s more intuitive.

**Comment by brianrourkeboll** [comment]
Implemented in #917, #919, #920.
